nucypher/utilities/networking.py

```diff
@@ -200,4 +204,62 @@ def determine_external_ip_address(
     # complete failure!
     if not rest_host:
         raise UnknownIPAddress('External IP address detection failed')
+
     return rest_host
+
+
+def _is_global_ipv4(ip: str) -> bool:
+    """Check if an IP address is a global IPv4 address according to RFC 1918."""
+    try:
+        ip = ip_address(ip.strip())
+        return isinstance(ip, IPv4Address) and ip.is_global
+    except AddressValueError:
+        return False
+
+
+def _resolve_ipv4(ip: str) -> Optional[str]:
+    """Resolve an IP address to IPv4 if required and possible."""
+    try:
+        ip = ip_address(ip.strip())
+    except AddressValueError:
+        return None
+    if isinstance(ip, IPv6Address) and ip.ipv4_mapped:
+        return str(ip.ipv4_mapped)
+    elif isinstance(ip, IPv4Address):
+        return str(ip)
+
+
+def _extract_ip(header: str, request: Request) -> Optional[str]:
+    """Extract the first IP address from a request header."""
+    if header in request.headers:
+        for ip in request.headers[header].split(","):
+            yield ip
+
+
+def _ip_sources(request: Request) -> str:
+    """Iterate over all possible sources of IP addresses in a request's headers."""
+    for header in ["X-Forwarded-For", "X-Real-IP"]:
+        yield from _extract_ip(header, request)
+    yield request.remote_addr
+
+
+def get_request_global_ipv4(request: Request) -> Optional[str]:
+    """
+    If the request is forwarded from a proxy, the first global IP address in the chain is returned.
+    'X-Forwarded-For' (XFF) https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Forwarded-For#selecting_an_ip_address
+
+    If XFF is not present in the request headers, this method also checks for 'X-Real-IP',
+    a popular non-standard header conventionally configured in some proxies like nginx.
+
+    Finally, if neither XFF nor X-Real-IP are present, the request is assumed to be
+    direct and the remote address is returned.
+
+    In all cases, tests that the IPv4 range is global. RFC 1918 privately designated
+    address ranges must not be returned.
+    https://www.ietf.org/rfc/rfc1918.txt
+
+    """
+    for ip_str in _ip_sources(request=request):
+        ipv4_address = _resolve_ipv4(ip_str)
+        if ipv4_address and _is_global_ipv4(ipv4_address):
+            return ipv4_address
```

Comment:
Does it make sense for `_resolve_ipv4` to instead return an `IPv4Address` object instead of a string? Then you don't really need to do the extra call to `ip_address(...)` in `_is_global_ipv4` which is duplicated and `_is_global_ipv4` is no longer needed. For example, you could end up just doing this:

```suggestion
    for ip_str in _ip_sources(request=request):
        ipv4_address = _resolve_ipv4(ip_str)
        if ipv4_address and ipv4_address.is_global:
            return str(ipv4_address)
```

wdyt?

Reply:
✔️ nice idea - I also think it's more pythonic to use higher level abstractions here.